### DISCUSSION TITLE
Sub Flows / Nested Flows

### DISCUSSION THREAD

**Discussion by moklick** [xyflow/xyflow#1024] [New Features]
Add a `children` option to nodes to enable sub flows.

**Comment by shushuitie2017** [comment]
It's good idea！

**Reply by pgupta56** [reply]
When to expect this feature in production??

**Comment by meatflavourdev** [comment]
@moklick I'm curious about this-- can you provide more detail about the idea-- how it would function and describe the intended use case?

**Reply by moklick** [reply]
The feature request came up here: #611. It's about the ability to render a flow within a node. It's like a group but that the group itself is a node with its own handles. In our app we don't have a use case for sub flows yet but depending on what you are building it could help to structure a flow.

**Comment by zijianjoy** [comment]
Sub flow is beneficial in terms of describing complex graph with layers. It is very similar to describing a flow with a function: We usually just write a few lines for a single function logic. When the function logic becomes complicated, we abstract it to another function, and just call this new function with a single line of function call. We keep each function concise in this way in programming languages. 

For react flow, having sub-flow is the same as the ability to call another function with a single line of function call. It strongly reduces complexity of a flow, and improves the visualization and readability of complex flows.

**Comment by neilpelow** [comment]
This feature would be very useful for a project I'm working on. I believe it would also require an extension of the drag & drop functionality? Assuming you would want to move nodes/sub-flows between different parents/containing nodes.

**Comment by neokike** [comment]
if anyone is interested, I was having this exact same requirement and what i ended up doing was creating a custom node whose content is just another ReactFlow instance, this allowed me to kind of mimic the subflow behavior.

![Captura de Pantalla 2021-09-14 a la(s) 15 04 43](https://user-images.githubusercontent.com/2183963/133272836-542003ab-3c1e-4d8c-9f39-210ef0a728a4.png)

 the ReactFlow custom node instance have all the features like panning, grabbing and zooming disabled but you can move the group around and i even added a resize handler to make the node bigger or smaller if needed.

i can probably make this more robust (like allowing drag and drop inside the group,  allow moving nodes inside or add the ability to dynamically create edges between the group nodes) but for my use case this was enough.

**Reply by zijianjoy** [reply]
That is awesome @neokike ! Would you like to share a demo repo which embeds a ReactFlow instance in a node?

**Reply by Sthikk** [reply]
@neokike - would you be able to share the custom node code for others who have been trying to solve this problem?  It would be a great help for the community.!

**Reply by gregorym** [reply]
@neokike can you share your custom node??

**Reply by cliffordfajardo** [reply]
@neokike - can you share a code snippet of the custom node? 🙏

**Comment by moklick** [comment]
We are currently working on the subflow/grouping feature. Which API do you prefer? 

use 🚀 reaction for the nested childNodes API: 

```js
const nodes = [
  { id: 'a', label: 'node-a' },
  { id: 'b', label: 'node-b' },
  { id: 'c', label: 'node-c', childNodes: [
    { id: 'd', label: 'child node-d' },
    { id: 'e', label: 'child node-e' },
  ]}
];
```

and 🎉  for the flat parentNode API: 

```js
const nodes = [
  { id: 'a', label: 'node-a' },
  { id: 'b', label: 'node-b' },
  { id: 'c', label: 'node-c' },
  { id: 'd', label: 'child node-d', parentNode: 'c' },
  { id: 'e', label: 'child node-e', parentNode: 'c' }
];
```

**Reply by meatflavourdev** [reply]
🙋‍♂️ nested please 😅

**Reply by JDvorak** [reply]
seconding nested format

**Reply by gilneto8** [reply]
I'll take the nested for sure.

**Reply by truong-hua** [reply]
The second one is easier for anyone to form a tree data structure as it's also the popular way that people usually use to store a tree in database or array. The latter one will help prevent an extra logic o convert from popular tree data structure to react-flow nested nodes.

**Reply by datoslabs** [reply]
I agree with @truong-hua, flat parentNode API is easier for users of this library especially when moving a leaf node from a deeply nested branch to another.  Currently this library's use of `elements` array is flat, therefore, to handle moving a child node 3+ levels down/nested to another parent node 3+ levels down/nested will require multiple loops using the first approach but 1 loop with the 2nd flat approach.

**Reply by ncarrara** [reply]
What if you need to do an intersection at some point? you won't be able to do that with nested, so flat it is for me. Also it follows the edge logic (flat).

**Reply by datoslabs** [reply]
I am hoping to convince some of you to change your vote with a quick comparison of the code involved (from the user's POV) in inserting a new `node 'f'` under `node 'e'` as its parent between the 2 approaches:

**Nested:**
```
const traverseAndUpdateTree = (parent, lookupFn, updateFn)=>{
	let updated = false;
	const updatedChildNodes = parent.childNodes.map(c=>{
		if (lookupFn(c)) {
			updated = true;
			return updateFn(c);
		}
		if (c.childNodes) {
			const [_c, _updated] = traverseAndUpdateTree(c, lookupFn, updateFn);
			if (_updated){
				updated = true;
				return _c;
			}
		}
		return c;
	});
}

setElements(els=>{
	const newNode = {id:'f', label:'grand child node-f'};
	const lookupFn = (n)=>n.id === 'e';
	const updateFn = (n)=>{
		const childNodes = n.childNodes ? n.childNodes.concat(newNode) : [newNode];
		return {...n, childNodes};
	};
	const updated = els.map(n=>{
		if (isEdge(n) || (n.id !== 'e' && (!n.childNodes || n.childNodes.length === 0))) 
			return n;
		if (n.id === 'e') {
			return updateFn(n);
		} else {
			return traverseAndUpdateTree(n, lookupFn, updateFn)[0];
		}
	});
	return updated; 
});
```

**Flat:**
```
setElements(els=>els.concat({id:'f', label:'grand child node-f', parentNode:'e'}));
```

Updating a node's parentage will be similar in terms of comparison.

**Reply by chrtze** [reply]
To see the differences and challenges in the implementation, we have roughly implemented both data structures with basic nested node behaviour:

First approach with nested childNodes: https://github.com/wbkd/react-flow/tree/nested-nodes
Second approach with flat node array and parentNode property: https://github.com/wbkd/react-flow/tree/nested-nodes-flat

Personally, I would prefer the flat version, because it is building on top of the already existing data structure instead of replacing the data structure to a nested object. That makes it easier to implement and as @datoslabs pointed out, way easier to manipulate the nodes from a user perspective.

In the nested version, we would have to add helper functions that allow mapping, iterating, or filtering (something like this: https://github.com/wbkd/react-flow/blob/nested-nodes/src/utils/nodes.ts) the nested tree recursively, whereas in the flat version we can use the existing array methods. One problem with the flat option is that we need to iterate over all nodes for finding all child nodes, whereas in the nested structure every node already knows their children.

<img width="746" alt="Screenshot 2021-10-21 at 17 52 59" src="https://user-images.githubusercontent.com/3797215/138313998-3102309a-daf6-4af0-bdba-9b0b2bf22524.png">

What do you think of the implementation? Any other suggestions or ideas on how to handle this?

**Reply by datoslabs** [reply]
This looks awesome!!  If I am not mistaken, the parent nodes' position and dimension in the current implementation are dynamic and cannot be specified by the user; if this is the case, do you plan to accept input from the user?  This is important because I have the requirement to style the parent/container node to add labels/headers and also to allow users to resize the container by dragging its corners (or edges).

**Reply by pdubey84** [reply]
+1 for flat and thanks a lot for picking this up.

**Reply by hartra344** [reply]
+1 for flat, way easier to work with

**Reply by truong-hua** [reply]
@chrtze I think that if you would like to optimize for a certain logic, we will have to go deep dive into algorithm and data structure. The data structure can be converted or indexed depend on different purpose in case of we would like to optimize for a single one operation.

But if the question is which one better for definition or declaration, I strongly recommend the flat as it is also the way of other storage components are doing and people can easily find it out in many hello world books.

If the question is which one is better for performance, we will need to fingure out the operation we are talking about and the trade off accepted. As the nested one will make you a little bit difficult to count the number of node. No data structure can fit with any operation.

Regarding definition purpose, the **FLAT** one will help to prevent unlimited cycle of node relations that may wrongly do by developer and save us a lot of time for validating it to prevent infinite loops.

**Reply by chrtze** [reply]
@truong-hua , good point. We also thought about converting the data structure to a indexed format or similar. The problems with that is that in some places in the code we benefit from iterating/filtering, in other places we want to target/find a specific element. As we want to keep it as simple as possible, in the best scenario we can use the "user" data structure equally internally without converting forth and back.

**Reply by chrtze** [reply]
> This looks awesome!! If I am not mistaken, the parent nodes' position and dimension in the current implementation are dynamic and cannot be specified by the user; if this is the case, do you plan to accept input from the user? This is important because I have the requirement to style the parent/container node to add labels/headers and also to allow users to resize the container by dragging its corners (or edges).

@datoslabs thanks for the input. This is more like a prototype, we are figuring out what behaviour we want to support in the new version and for the subgraphs. So, your feedback is very valuable here as we are not so sure, which use-cases exist for subgraphs and how we can support them.

**Reply by datoslabs** [reply]
Is there a thread where we are tracking/grooming the use-cases for this feature?  

Also, have you explored the option of introducing a new `container` node type and render children nodes as nested `divs` under it?  I took a similar approach as you did treating the `container`  node like any other node and maintaining a dual data structure (flat + nested); I didn't realize the amount of work needed to allow custom styling of the `container` node until I went down the rabbit hole and gave me a renewed appreciation for CSS.  There are a few use-cases that I think are important from a UI/UX standpoint, most of which will be made easier if CSS can take over the layout work using nested `divs`.

**Reply by chrtze** [reply]
> Is there a thread where we are tracking/grooming the use-cases for this feature?

I think some things are already in the issue and some in this discussion, but a list would be helpful too.

> Also, have you explored the option of introducing a new `container` node type and render children nodes as nested `divs` under it? I took a similar approach as you did treating the `container` node like any other node and maintaining a dual data structure (flat + nested); I didn't realize the amount of work needed to allow custom styling of the `container` node until I went down the rabbit hole and gave me a renewed appreciation for CSS. There are a few use-cases that I think are important from a UI/UX standpoint, most of which will be made easier if CSS can take over the layout work using nested `divs`.

Yes, I thought about that too. It would be saving some iterations on the nodes when dragging the parent node, because we would not have to move all child nodes accordingly. I will give it a try next week!

**Reply by vilasone455** [reply]
+1 for flat format

**Comment by moklick** [comment]
While implementing this feature we are exploring some other challenges. Right now a parent node receives the dimension and position based on the position of the children. The children are in the same coordinate system as their parents. This makes it hard position the parent since it always depends on its children. The advantage is that you only need to deal with one coordinate system. Another solution would be to give a parent node its own dimensions and coordinates and position the children relative. What do you think about it?

**Reply by neo7337** [reply]
The only advantage I see on keeping the parent node and children positions and dimensions relative is that you might get fine grained control on each flow/nested flow that we might end up building

**Reply by datoslabs** [reply]
@moklick, I believe option II (relative) is more flexible and simpler to implement in the end.  

When I implemented my own, the children nodes are placed in the same coordinate system (because I didn't want to modify the core React-Flow lib) and I found myself having to write code to handle the layout and positioning of parent-children nodes which are normally handled by CSS.  For example, to correctly calculate the size of the parent node with a label (say `<span> Parent Container </span>`), not only I needed to know the size and position of its children nodes + any children nodes' CSS styling not accounted by `getBoundingClientRect()`, I also needed to know the size and position of the parent label ( and CSS styling affecting label's visual size/positioning), in order to correctly size the parent node.  This is ultimately the reason I stopped my custom nested node implementation and suggested placing children nodes as nested `divs` under their parent nodes in my response above to @chrtze.  When children nodes are positioned relative to the parent (an option when it's implemented as nested `divs`), it provides a lot of flexibility by delegating common layout/styling functionality (e.g. margin, padding, block/flex, etc) to CSS.

**Reply by pgupta56** [reply]
@moklick when is v10 releasing and subflow is available??

**Reply by moklick** [reply]
hopefully we can ship a beta by end of November

**Reply by pgupta56** [reply]
Thanks @moklick !! great work!!

**Reply by neo7337** [reply]
> 

When are we able to test the beta @moklick ?

**Reply by moklick** [reply]
@neo7337 You can find the instructions here https://github.com/wbkd/react-flow/pull/1555

**Comment by neo7337** [comment]
@moklick I am currently trying to implement the nested flow in custom node, I will appreciate any help if it can achieved with the current framework. I know it would be complex, but if there is some way around in the current framework.
Looking for some help!

**Reply by datoslabs** [reply]
@neo7337, it can be implemented as a custom node; I have done it and there are several other examples of other implications shared in #1022.  The branches nested-nodes and nested-nodes-flat also contain useful code to help implement nested flow in a custom node.

**Reply by neo7337** [reply]
@datoslabs can you please share the sample code of this example, that would really help me get my stuff going
https://github.com/wbkd/react-flow/discussions/1022#discussioncomment-1504321
Thanks!!

**Reply by datoslabs** [reply]
I think it will be much faster and easier to use the changes @chrtze made under nested-nodes and nested-nodes-flat branches (take a look at the previous replies by him in this discussion, he included screenshots) as a starting point.  His code is much cleaner and better structured, thus a better starting point to add resizability and DnD children nodes into parent nodes functionality (if you need these features).

What I had implemented was for evaluation/demo purposes and very messy; it was clobbered together over several attempts with varying requirements (e.g. user draggable handles/pseudo handles, polymorphic nodes, etc.) that are not related to nested flows so it would take some time to clean it up.

**Reply by chrtz** [reply]
@chrtze - i think you earned some respect here. :)

**Comment by moklick** [comment]
👉 If you want to test the new subflow feature, you can check the instructions here: #1555

**Reply by neo7337** [reply]
Will you be supporting the drag and drop feature for nested flow as well @moklick ?

**Reply by datoslabs** [reply]
Based on my review of the PR, if I understood correctly, you should be able to implement the DnD features for nested flows using the `onMouseEnter | onMouseMove | onMouseLeave` event handlers.

**Comment by datoslabs** [comment]
@moklick and team, thank you again for all the great work!  I was only able to spend a brief moment reviewing the PR and am encouraged that the core functionality to implement nested flow seems to be coming together nicely.  

There does not appear to be a dedicated "container" node type so far; I am wondering if you plan to add it as part of V10 release - I know you have already done a lot of work and V10 schedule has been pushed back :-)  If not, I don't think it's a big deal as the core features are already in place to implement one, it can be added as an example (I can help with a PR over the holidays if the V10 branch is near finished).

**Reply by datoslabs** [reply]
Hi @chrtze, As I started thinking about an example, if I would like to customize the layout of parent nodes like making it resizable with say `react-resizable` so that children nodes are rendered inside `<Resizable></Resizable>`, how would I be able to achieve this?

Update: I am not sure if I understood correctly but based on the code in V10 branch it appears that parent-children `nodes` are not in nested `divs` which means adding scrolls and/or using  `<Resizable>` to expand/contract parent viewable areas will not be easily achievable, is this correct?

**Reply by moklick** [reply]
We think that it is more flexible when we don't introduce a new type. You should be able to create your own `group` node type and use `react-resizable` for that. Does that work for you? (We tried to nest the divs but it got very complicated and produced a bunch of other issues regarding zIndices and positioning.) 

For now the main sub flow features are:
* children get dragged when the parent gets dragged
* children get positioned relative which makes it easier to layout them
* you can set `nodeExtent="parent"` for a child node so that one can not drag it outside of its parent node 
* when an edge goes inside a group it is on top of the other nodes

I am sure we will add more sub flow features when v10 is released.

**Reply by datoslabs** [reply]
Sounds like it's similar to my current implementation.  Is there a way user can render children nodes in their custom nodes?  React- resizable without nested `divs` is very limited; I have it in my current implementation and since the node `divs` are not nested, I was unable to make it overflow with scroll bars and needed to constantly check the children's collective bounding box to make sure it doesn't visually "orphan" any children during resizing.  Styling the content of the parent node (like title) while trying to maintain margin/padding relative to its children without `flex box` is very cumbersome as well.

**Comment by AlaaHamidah** [comment]
![Screenshot 2021-12-07 215304](https://user-images.githubusercontent.com/26914921/145136647-90228261-a91e-46f0-9c58-6c0a24a5f136.jpg)

Hi, 
Can I achieve something like this with sub-flow?
 I'm not sure what to call it .. it's not a sub-flow, it's kind of multiple flows connected together

**Reply by moklick** [reply]
Hey, that should be possible!

**Reply by pgupta56** [reply]
Great !! It is what we are also looking at! Thanks a zillion for making out work easier !! :-)

**Comment by nomis-dev** [comment]
Is it possible to adapt the parent node to the size of subflow?

**Reply by chrtze** [reply]
No, this is not possible right now. The easiest way to do it is probably to calculate the bounding box of the child nodes in the `onNodesChange` handler and size the parent accordingly. We are thinking of adding a hook/helper method to make this easier in the future.

**Comment by moklick** [comment]
This feature is available since v10 https://reactflow.dev/docs/guides/sub-flows/
